src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,231 @@
+      weight_char_field<Tags::VPsi>(v_psi_ext, get<0>(char_speeds_ext), 1.);
+  get<Tags::VZero<Dim>>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VZero<Dim>>(v_zero_ext, get<1>(char_speeds_ext),
+                                          1.);
+  get<Tags::VPlus>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VPlus>(v_plus_ext, get<2>(char_speeds_ext), 1.);
+  get<Tags::VMinus>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VMinus>(v_minus_ext, get<3>(char_speeds_ext), 1.);
+}
+}  // namespace CurvedScalarWave::BoundaryCorrections::CurvedScalarWave_detail
+
+namespace CurvedScalarWave::BoundaryCorrections {
+template <size_t Dim>
+UpwindPenalty<Dim>::UpwindPenalty(CkMigrateMessage* msg) noexcept
+    : BoundaryCorrection<Dim>(msg) {}
+
+template <size_t Dim>
+std::unique_ptr<BoundaryCorrection<Dim>> UpwindPenalty<Dim>::get_clone()
+    const noexcept {
+  return std::make_unique<UpwindPenalty>(*this);
+}
+
+template <size_t Dim>
+void UpwindPenalty<Dim>::pup(PUP::er& p) {
+  BoundaryCorrection<Dim>::pup(p);
+}
+
+template <size_t Dim>
+double UpwindPenalty<Dim>::dg_package_data(
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_psi,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_v_zero,
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_plus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_minus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_gamma2,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_interface_unit_normal,
+    const gsl::not_null<tnsr::a<DataVector, 3, Frame::Inertial>*>
+        packaged_char_speeds,
+
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const Scalar<DataVector>& psi,
+
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& shift,
+    const tnsr::II<DataVector, Dim, Frame::Inertial>& inverse_spatial_metric,
+    const Scalar<DataVector>& constraint_gamma1,
+    const Scalar<DataVector>& constraint_gamma2,
+
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity)
+    const noexcept {
+  *packaged_gamma2 = constraint_gamma2;
+  *packaged_interface_unit_normal = interface_unit_normal;
+
+  {  // package characteristic fields
+    const auto char_fields =
+        characteristic_fields(constraint_gamma2, inverse_spatial_metric, psi,
+                              pi, phi, interface_unit_normal);
+    *packaged_v_psi = get<Tags::VPsi>(char_fields);
+    *packaged_v_zero = get<Tags::VZero<Dim>>(char_fields);
+    *packaged_v_plus = get<Tags::VPlus>(char_fields);
+    *packaged_v_minus = get<Tags::VMinus>(char_fields);
+  }
+
+  {  // package characteristic speeds
+    Scalar<DataVector> shift_dot_normal{};
```

Comment:
I would be happy to make this change but I worry this might not play well with other parts of the code, as all evolution systems at the moment use `std::array`'s. A quick grep tells me char speeds are directly used in, for e.g., the fluxes:

```
NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/LocalLaxFriedrichs.hpp
NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/Hll.hpp
```

@nilsdeppe can you please confirm that swapping out the type of characteristic speeds from `std::array<DataVector, 4> --> tnsr::a<3>` not break anything else?

Reply:
Those fluxes aren't used anymore by the evolution code (I don't think by the elliptic code either). However, since every other system's `characteristic_speeds` function uses `std::array`s it seems it would be better not to change it in just one system. It is possible to assign a `std::array` to a `tnsr::a<3>`, but it might have to be done explicitly: `*packaged_char_speeds = tnsr::a<3, DataVector, Frame::Inertial>{characteristic_speeds(...)};`

Reply:
I agree that this should change simultaneously in all evolution systems. And that is why I'd rather not call the `CurvedScalarWave::characteristic_speeds` function in `dg_package_data` as that would need an additional allocation of a `std::array<DV, 4>`. Its true this leads to code duplication, but I can't think of a better solution given the constraints.

Reply:
ok, I would argue that avoiding this code duplication is more important than keeping a completely unused function consistent with other unused functions, which are now essentially deprecated. Keep in mind that I will have to duplicate it again for the boundary conditions as will pretty much any other boundary condition / boundary correction implementation.

But if you decide to leave it, I can port this over for all systems before I do the boundary conditions - I dont want to hold this PR up.

Reply:
Okay, I am convinced. As a middle ground, I have added an overload of `characteristic_speeds` that returns a `tnsr::a<3>` by reference instead of a `std::array`. I have not deleted the version(s) that return(s) `array`s as that should be done in a different PR for all evolution systems simultaneously. 

This overload solves our purpose in both the flux and boundary conditions for this system.

Reply:
I think that is a good solution